File: sort.py
def compareSwap(arr, a, b):
    if arr[a] < arr[b]:
        tmp = arr[a]
        arr[a] = arr[b]
        arr[b] = tmp
        return True
    return False

def bubbleSort(arr):
    max = len(arr) - 1
    while max > 0:
        swapped = False
        for i in range(1, max+1):
            if compareSwap(arr, i, i-1):
                swapped = True
        if not swapped:
            break
        max -= 1
    return arr

def shakerSort(arr):
    min = 0
    max = len(arr) - 1
    for t in range(max+1):
        swapped = False
        if t%2 == 0:
            for i in range(min+1, max+1):
                if compareSwap(arr, i, i-1):
                    swapped = True
            max -= 1
        else:
            for i in range(max-1, min-1, -1):
                if compareSwap(arr, i+1, i):
                    swapped = True
            min += 1
        if not swapped:
            break
    return arr

def selectionSort(arr):
    for i in range(len(arr)):
        idx = i
        for c in range(i, len(arr)):
            if arr[c] < arr[idx]:
                idx = c
        if idx != i:
            tmp = arr[idx]
            arr[idx] = arr[i]
            arr[i] = tmp
    return arr

File: sort.rb
class Sort
  def self.bubble(arr)
    (arr.length - 1).downto(1).each do |last|
      swapped = false
      (1..last).each do |i|
        if arr[i] < arr[i-1]
          smaller = arr[i]
          arr[i] = arr[i-1]
          arr[i-1] = smaller 
          swapped = true
        end
      end
      break unless swapped
    end
    arr
  end

  def self.shaker(arr)
    min = 0
    max = arr.length-1
    (arr.length-1).times do |t|
      swapped = false
      interval = if t.even? then (min+1).upto(max) else (max-1).downto(min) end
      interval.each do |i|
        a = if t.even? then i-1 else i end
        b = if t.even? then i else i+1 end
        if arr[a] > arr[b]
          tmp = arr[a]
          arr[a] = arr[b]
          arr[b] = tmp
          swapped = true
        end
      end
      if t.even? then (max -= 1) else (min += 1) end
      break unless swapped
    end
    arr
  end

  def self.selection(arr)
    arr.length.times do |i|
      smallest_i = i
      smallest = arr[i]
      (i...arr.length).each do |c|
        if arr[c] < smallest
          smallest_i = c
          smallest = arr[c]
        end
      end

      unless smallest_i == i
        arr[smallest_i] = arr[i]
        arr[i] = smallest
      end
    end
    arr
  end
end
